java/integTest/src/test/java/glide/standalone/PubSubTests.java

```diff
@@ -0,0 +1,827 @@
+import lombok.SneakyThrows;
+import org.apache.commons.lang3.tuple.Pair;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+// @Timeout(30) // sec
+public class PubSubTests {
+
+    // TODO protocol version
+    @SneakyThrows
+    @SuppressWarnings("unchecked")
+    private <M extends ChannelMode> BaseClient createClientWithSubscriptions(
+            boolean standalone,
+            Map<M, Set<String>> subscriptions,
+            Optional<MessageCallback> callback,
+            Optional<Object> context) {
+        if (standalone) {
+            var subConfigBuilder =
+                    StandaloneSubscriptionConfiguration.builder()
+                            .subscriptions((Map<PubSubChannelMode, Set<String>>) subscriptions);
+
+            if (callback.isPresent()) {
+                subConfigBuilder.callback(callback.get(), context.get());
+            }
+            return RedisClient.CreateClient(
+                            commonClientConfig().subscriptionConfiguration(subConfigBuilder.build()).build())
+                    .get();
+        } else {
+            var subConfigBuilder =
+                    ClusterSubscriptionConfiguration.builder()
+                            .subscriptions((Map<PubSubClusterChannelMode, Set<String>>) subscriptions);
+
+            if (callback.isPresent()) {
+                subConfigBuilder.callback(callback.get(), context.get());
+            }
+
+            return RedisClusterClient.CreateClient(
+                            commonClusterClientConfig()
+                                    .subscriptionConfiguration(subConfigBuilder.build())
+                                    .build())
+                    .get();
+        }
+    }
+
+    private <M extends ChannelMode> BaseClient createClientWithSubscriptions(
+            boolean standalone, Map<M, Set<String>> subscriptions) {
+        return createClientWithSubscriptions(
+                standalone, subscriptions, Optional.empty(), Optional.empty());
+    }
+
+    @SneakyThrows
+    private BaseClient createClient(boolean standalone) {
+        if (standalone) {
+            return RedisClient.CreateClient(commonClientConfig().build()).get();
+        }
+        return RedisClusterClient.CreateClient(commonClusterClientConfig().build()).get();
+    }
+
+    /** Message queue used in callback to analyze received messages. Number is a client ID. */
+    private final ConcurrentLinkedDeque<Pair<Integer, Message>> mq = new ConcurrentLinkedDeque<>();
+
+    /** Clients used in a test. */
+    private final List<BaseClient> clients = new ArrayList<>();
+
+    @BeforeEach
+    @SneakyThrows
+    public void cleanup() {
+        mq.clear();
+        for (var client : clients) {
+            if (client instanceof RedisClusterClient) {
```

Comment:
If this test works for both standalone and cluster, we should move it up one directory.